Refactor the writing of result files

milljm opened the pull request:
Concerning TestHarness features:
  --sep-files
  --sep-files-ok
  --sep-files-fail
  --output-dir

Consolidate the writing of anything 'resulty' which was
occurring at the end of a TestHarness run_tests run.

Store and work with a json file for failed tests (actually
all tests). This will allow anyone to easily parse the results
of their tests for what ever means they need it for.

The results json file is also a stepping stone towards doing
away with a PBS session file.

Closes #11116

Commit:
Refactor the writing of result files

Concerning TestHarness features:
  --sep-files
  --sep-files-ok
  --sep-files-fail
  --output-dir

Consolidate the writing of anything 'resulty' which was
occurring at the end of a TestHarness run_tests run.

Store and work with a json file for failed tests (actually
all tests). This will allow anyone to easily parse the results
of their tests for what ever means they need it for.

The results json file is also a stepping stone towards doing
away with a PBS session file.

Closes #11116

Commit:
Fix unused variables for assert tests

Closes #11116

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/170557/) on 5fb83c2 wanted to post the following:

View the site [here](http://mooseframework.inl.gov/docs/PRs/11123/site)

This comment will be updated on new commits.

Commit:
Remove the ability to save output

Do not save output and params for every test. This functionality
is not necessary.

Closes #11116

Commit:
Fix sep-files exception issue

Address comments and fix issues with --sep-files.
Refactor the logic a bit when using said option.

Do not save information about silent tests (--re
generates a lot of these).

Closes #11116

Commit:
Add entry to make detecting failures easy

Instead of trying to figure out by status name, if
a test has failed, store a bool instead.

Closes #11116
